### ISSUE TITLE
some problems when use "docker stats" with kata-runtime

### ISSUE THREAD

**Issue by liangxianlong** [moby/moby#38329]
* [x] This is a bug report
* [x ] This is a feature request
* [ ] I searched existing issues before opening this one

### Expected behavior
In this case, the runtime is kata-runtime, when use "docker stats ${containerid}", the output about CPU% make me confused。We can see the  calculations in docker as follows(stats_helpers.go):
```
func calculateCPUPercentUnix(previousCPU, previousSystem uint64, v *types.StatsJSON) float64 {
	var (
		cpuPercent = 0.0
		// calculate the change for the cpu usage of the container in between readings
		cpuDelta = float64(v.CPUStats.CPUUsage.TotalUsage) - float64(previousCPU)
		// calculate the change for the entire system between readings
		systemDelta = float64(v.CPUStats.SystemUsage) - float64(previousSystem)
		onlineCPUs  = float64(v.CPUStats.OnlineCPUs)
	)

	if onlineCPUs == 0.0 {
		onlineCPUs = float64(len(v.CPUStats.CPUUsage.PercpuUsage))
	}
	if systemDelta > 0.0 && cpuDelta > 0.0 {
		cpuPercent = (cpuDelta / systemDelta) * onlineCPUs * 100.0
	}
	return cpuPercent
}
```
The ```v.CPUStats.CPUUsage.TotalUsage``` and ```previousCPU``` are obtained by kata-runtime from the guest or virtual machine. 
The ```v.CPUStats.SystemUsage``` ,```previousSystem``` and ```onlineCPUs``` are obtained by dockerd from the host, the code we can see in docker(stats_collector_unix.go) as flow:
```
func (s *statsCollector) getSystemCPUUsage() (uint64, error) {
	var line string
	f, err := os.Open("/proc/stat")
	if err != nil {
		return 0, err
	}
	defer func() {
		s.bufReader.Reset(nil)
		f.Close()
	}()
	s.bufReader.Reset(f)
	err = nil
	for err == nil {
		line, err = s.bufReader.ReadString('\n')
		if err != nil {
			break
		}
		parts := strings.Fields(line)
		switch parts[0] {
		case "cpu":
			if len(parts) < 8 {
				return 0, fmt.Errorf("invalid number of cpu fields")
			}
			var totalClockTicks uint64
			for _, i := range parts[1:8] {
				v, err := strconv.ParseUint(i, 10, 64)
				if err != nil {
					return 0, fmt.Errorf("Unable to convert value %s to int: %s", i, err)
				}
				totalClockTicks += v
			}
			return (totalClockTicks * nanoSecondsPerSecond) /
				s.clockTicksPerSecond, nil
		}
	}
	return 0, fmt.Errorf("invalid stat format. Error trying to parse the '/proc/stat' file")
}

func (s *statsCollector) getNumberOnlineCPUs() (uint32, error) {
	i, err := C.sysconf(C._SC_NPROCESSORS_ONLN)
	if err != nil {
		return 0, err
	}
	return uint32(i), nil
}
```
**This calculation is fine for runc, but it is debatable for kata-runtime.**

**Output of `docker version`:**

```
Client:
 Version:      17.03.2-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   6ab216c
 Built:        Fri Jul  6 06:08:57 2018
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.2-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   6ab216c
 Built:        Fri Jul  6 06:08:57 2018
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 109
 Running: 2
 Paused: 0
 Stopped: 107
Images: 1
Server Version: 17.03.2-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc kata-runtime
Default Runtime: kata-runtime
Init Binary: docker-init
containerd version: 6ab216c7a67dad4b4b99cdba3c1d5f6512bfbd9e (expected: 4ab9917febca54791c5f071a9d1f404867857fcc)
runc version: 6ab216c7a67dad4b4b99cdba3c1d5f6512bfbd9e (expected: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe)
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.21.1.el7.x86_64
Operating System: NewStart Carrier Grade Server Linux Core 5.04
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 78.39 GiB
Name: host-97
ID: 367P:DCJM:2WMT:XURG:BZ2O:JU64:M5IX:JYWD:7WV5:DHE6:RIS4:CRQ7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 26
 Goroutines: 31
 System Time: 2018-12-07T09:01:22.027455904+08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**
The kata-containers link:
[https://github.com/kata-containers](url)
@GordonTheTurtle @evanphx @uggedal @kevinclark

**Comment by evanphx** [comment]
Why did you tag me?

**Comment by olljanat** [comment]
@liangxianlong 17.03 is very old and unsecure version so plz upgrade to latest version (18.09) first and check if behaviour is same on it.

**Comment by olljanat** [comment]
> In this case, the runtime is kata-runtime, when use "docker stats ${containerid}", the output about CPU% make me confused。

@liangxianlong can you also share example output and tell what is confusing on it?

**Comment by liangxianlong** [comment]
Because，we need docker-ce-18.06. As flow:
1.first case in runc:
```
# docker run -ti -d --runtime runc busybox /bin/sh
a321becf59f89f82fa999784eca2b62c96ac03c034be34ae06462a8cd0c6a0cc

# docker stats a321becf59f89f82fa999784eca2b62c96ac03c034be34ae06462a8cd0c6a0cc
CONTAINER ID        NAME                  CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
a321becf59f8        infallible_stallman   0.00%               52KiB / 3.026GiB    0.00%               438B / 0B           897kB / 0B          1
```
2.second case in kata-runtime, the CPU % and MEM USAGE, NET I/O, BLOCK I/O is not correct.
```
# docker run -ti -d --runtime kata-runtime busybox /bin/sh
03832f2b78922529019fdf7d73ccfb9bcdb00e746e7ea1088f175d1276925f9b

# docker stats 03832f2b78922529019fdf7d73ccfb9bcdb00e746e7ea1088f175d1276925f9b
CONTAINER ID        NAME                 CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
03832f2b7892        optimistic_volhard   0.54%               207.1MiB / 3.026GiB   6.68%               438B / 796B         6.04MB / 18.2MB     100
```

**Comment by liangxianlong** [comment]
But i don't think so. Because, docker stats uses OCI events interface. In runc, it reads host cgroup.in kata-runtime,it reads guest cgroup.
@olljanat

**Comment by liangxianlong** [comment]
Hi, I have found the cause of this difference. In docker-ce-17.03,"docker stats" use OCI events interface to get values;but in docker-ce-18.06,it does not. I think ,this is a real bug!@olljanat@thaJeztah

**Comment by olljanat** [comment]
Derek add label: area/runtime

**Comment by thaJeztah** [comment]
Docker 17.11 and up use containerd 1.x, and on those versions stats collecting is handled by containerd. If there's a change to be made, I guess this has to be done in containerd.

**Comment by thaJeztah** [comment]
I see you opened a ticket in the containerd issue-tracker; https://github.com/containerd/containerd/issues/2906

**Comment by liangxianlong** [comment]
Yes，I think so.  But i haven't get any response.Thinks

**Comment by liangxianlong** [comment]
Hi, for this question, containerd community merges this PR [https://github.com/containerd/containerd/pull/2434](url), and kata community merges this PR [https://github.com/kata-containers/runtime/pull/572](url), But the docker community has not made the corresponding changes at present.
 @thaJeztah

**Comment by thaJeztah** [comment]
Have you tried on Docker 18.09 ? The Runtime V2 (shim API) was released in containerd 1.2 (https://github.com/containerd/containerd/releases/tag/v1.2.0). Containerd 1.2 was released _after_ Docker 18.09 (which uses containerd 1.1.x), but Docker 18.09 ships with containerd 1.2.x

**Comment by liangxianlong** [comment]
Yes,I have tried ```docker-ce-18.09``` and ```ctr```.There are some of my attempts：
1. use ```ctr```
```
# ctr images ls
REF                               TYPE                                                 DIGEST                                                                  SIZE      PLATFORMS   LABELS 
10.74.121.41:5000/nginx/v1:latest application/vnd.docker.distribution.manifest.v2+json sha256:c235c3a31ed94016201b04e5ff8ac7cf8d8eca75811663e15ba7893514f142ef 116.4 MiB linux/amd64 -
```
```
# ps axf
3742 ?        Ssl   11:27 /usr/local/bin/containerd
9925 ?        Ssl    0:44 /usr/bin/dockerd -H fd://
```
```
# ctr run -d -t --runtime io.containerd.kata.v2 10.74.121.41:5000/nginx/v1:latest test
```
```
# ctr t ls
TASK    PID      STATUS    
test    15125    RUNNING
```
```
# ctr t metric test
ID      TIMESTAMP                                  
test    2019-01-10 23:14:07.589304061 +0000 UTC    
METRIC                   VALUE                       
memory.usage_in_bytes    3776512                     
memory.limit_in_bytes    9223372036854771712         
memory.stat.cache        0                           
cpuacct.usage            43392771                    
cpuacct.usage_percpu     [43392771 0 0 0 0 0 0 0]    
pids.current             1                           
pids.limit
```
But when i use docker-ce-18.09,i don't know how to specify shim-v2 because docker does not provide similar parameters.
